vllm/attention/backends/flash_attn.py

```diff
@@ -708,7 +714,9 @@ def unified_flash_attention(
         # Decoding run.
         # Use flash_attn_varlen_func kernel for speculative decoding
         # because different queries might have different lengths.
+
         assert decode_meta.max_decode_query_len is not None
+        # use only for actual varlen decoding
```

Comment:
To clarify my understanding of why the corruption issue was not showing up in head/other use cases --

The problem only happens when you use  flash_attn_varlen_func (which uses query_start_loc) in the prefill + decode use case. This code is only triggered currently by sd + chunked_prefill and so it showed up here. In the vanilla chunked_prefill case we will using the kv cache based implementation (flash_attn_with_kvcache) which does not use query_start_loc and hence the problem was not showing up. Is my understanding correct?

Reply:
Exactly! In particular the `if decode_meta.max_decode_query_len > 1` made sure it only triggered for decode batches coming from the `mqa_scorer` (that's also why the BatchExpansion scorer has been working fine all along)